### PR TITLE
feat(TableContainer): support id-ref for the header

### LINKED ISSUE #3805: DataTable TableContainer title/description cannot be referenced to label/describe Table (via aria-labelledby/aria-describedby)
TableContainer does not define IDs for its title and description nodes so that they can be used to label/describe the Table (via aria-labelledby/aria-describedby). 

Because of this, tables must either be labeled by title + description using a single ID on the TableContainer:
image

Or tables must use aria-label for labeling (which is not recommended by ARIA - "If the label text is visible on screen, authors SHOULD use aria-labelledby and SHOULD NOT use aria-label. " -https://www.w3.org/TR/wai-aria-1.1/#aria-label). This also makes it impossible to provide the screen reader a description for the table.

Without an aria-label or aria-labelledby attribute, JAWS uses the table's rows to label it:
image

## Environment

Firefox ESR 60.8.0esr
JAWS Version 2019.1906.10 ILM

## Detailed description

carbon-components@10.5.1
carbon-components-react@7.5.1

## Steps to reproduce the issue

https://codesandbox.io/s/cannot-connect-tablecontainer-titledescription-to-table-b19fm

## Additional information

- Screenshots or code
- Notes

### PR DESCRIPTION
**Pull request by asudoh**
This change adds a capability to specify the element ID of the `<h4>` in `<TableContainer>`. This allows application to set up our data table so VoiceOver announces the content in such `<h4>` when it announces "You are currently in a table" for `<table>`.

Fixes #3805.

#### Changelog

**New**

- `titleId` prop in `<TableContainer>`.

#### Testing / Reviewing

Testing should make sure `<TableContainer>` is not broken.

### CODE REVIEW

**Comment by abbeyhrt** [comment]
Hey! @asudoh since this adds a new prop, is this technically a feat, not a fix?

**Comment by asudoh** [comment]
@abbeyhrt You are right technically, I made it a `fix` given the addition is minor and it's for an a11y fix. But don't hesitate to speak up if you want `feat` commit message. Thanks!

**Comment by joshblack** [comment]
For reference, this falls under this section in our versioning guide: https://github.com/carbon-design-system/carbon/blob/master/docs/guides/versioning.md#a-new-prop-is-added-to-a-component

**Review by joshblack — COMMENTED** [review]
Definitely a bummer that we would need to add props to target inner children. I would assume we would also need to add `descriptionId` for `aria-describedby`? Are there any alternatives we could explore that don't rely on this?

**Comment by asudoh** [comment]
An alternative may be `aria-label={title}`. If we are OK with screen reader speaking both of title and description, no change is required for fixing the issue.

**Review by emyarod — APPROVED** [review]
this looks good to me

**Comment by joshblack** [comment]
@asudoh got it 👍 How would this work with the secondary label for the table btw? Specifically `bx--data-table-header__description`

**Comment by asudoh** [comment]
User of assistive technology can still hear the secondary label via a way of reading regular text.

**Comment by joshblack** [comment]
Would we need to use `aria-describedby` the same way we need `aria-labelledby`? That's all that I'm asking. I believe what you said also would apply to the title, but we are making this addition to the component API to support it. Specifically, the issue calls out that the title and description are unable to be used which is why I was asking about it.

**Comment by asudoh** [comment]
> Would we need to use `aria-describedby` the same way we need `aria-labelledby`?

Are you asking adding both of `aria-labelledby` pointing to title and `aria-describedby` pointing to description to `<Table>`, or something else?

**Comment by joshblack** [comment]
That was my interpretation from the linked issue, let me know if I'm off here! Just trying to understand requirements to figure out the best way to support them.

**Comment by asudoh** [comment]
Adding both of `aria-labelledby` pointing to title and `aria-describedby` pointing to description to `<Table>` causes VoiceOver read only the description not the title.

**Comment by joshblack** [comment]
@asudoh I'm unable to reproduce, could you help me understand what the issue is?

demo

Regardless, would it make sense for this to configurable as a prop for the user to specify for `title` and `description` or would it help to integrate this with `DataTable` and `getTableProps`? Maybe through context? Another option seems to be not passing in a `title` and rendering the markup as `children`, in this case, would we want to expose the title and description as components that people could use?

**Comment by asudoh** [comment]
I see both title/description are read on forward-navigation, but not backward-navigation.

**Comment by joshblack** [comment]
Oh no! What were the steps that ended up with you getting that behavior? Here is what I was trying:

demo

**Comment by asudoh** [comment]
I'm late on this, but here are the steps:

1. Checkout the branch in this PR
2. Run the React Storybook
3. Open http://localhost:9000/iframe.html?id=datatable--default
4. Go to DOM inspector
5. Add `id="table-container-description"` to `<p class="bx--data-table-header__description">`
6. Replace `<table aria-describedby="table-container-title" class="bx--data-table bx--data-table--no-border">` with `<table aria-labelledby="table-container-title" aria-describedby="table-container-description" class="bx--data-table bx--data-table--no-border">`
7. Set keyboard focus on "Show Info" button (The DOM element that comes next to the container of the `<div class="bx--data-table-container">`)
8. Turn on VoiceOver
9. Hit Ctrl-Option-Left
10. VO annnounces "end of table", then "With default options. You are currently in a table. To navigate the cells in this table press Control-Option, and then Up Arrow, Down Arrow, Left Arrow, or Right Arrow."

Leaving such problem is OK for me though. Wrt `getTableProps()`, are you thinking of auto-generating the ID-refs with it (as well as `getTableContainerProps()`)?

**Comment by joshblack** [comment]
I think we could definitely go down that route if we're doing the `titleId` and `descriptionId` for `aria-labelledby` and `aria-describedby` since we already have `getTableContainerProps`.

Another idea was to create `TableTitle` and `TableDescription` components and do the wiring up like you have currently, e.g.

```jsx
<TableContainer>
  <TableTitle id="table-title">Title of the table</TableTitle>
  <TableDescription id="table-desc">Table description</TableDescription>
  <Table aria-labelledby="table-title" aria-describedby="table-desc">
    {/* ... */}
  </Table>
</TableContainer>
```

**Comment by asudoh** [comment]
Back on this, and made a change to auto-generate ID refs for `aria-labeledby` and `aria-describedby`.

File: packages/react/src/components/DataTable/DataTable.js [hunk 1]
```diff
@@ -373,6 +373,8 @@ export default class DataTable extends React.Component {
       useStaticWidth,
       shouldShowBorder,
       stickyHeader,
+      'aria-labeledby': `${this.getTablePrefix()}__table-container__title`,
```
Comment (abbeyhrt) [inline packages/react/src/components/DataTable/DataTable.js]:
```suggestion
      'aria-labelledby': `${this.getTablePrefix()}__table-container__title`,
```
For some reason, ARIA uses the British English "labelled" not the American English "labeled" 🤷‍♀

Comment (asudoh) [reply packages/react/src/components/DataTable/DataTable.js]: Good catch - Thanks @abbeyhrt!

**Pull request closed without merging**